File: Python/DynamicProgramming/findTargetSumWays.py
def findTargetSumWays(nums, target):
    n = len(nums)
    summation = sum(nums)

    dp = [[None] * (2 * summation + 1) for _ in range(n)]

    def helper(index, sum_nums):
        # base case
        if index < 0:
            return 1 if sum_nums == target else 0

        if dp[index][sum_nums + summation] is not None:
            return dp[index][sum_nums + summation]

        # choose -nums[index]
        negative = helper(index - 1, sum_nums - nums[index])

        # choose +nums[index]
        positive = helper(index - 1, sum_nums + nums[index])

        dp[index][sum_nums + summation] = negative + positive
        return dp[index][sum_nums + summation]

    return helper(n - 1, 0)

File: Golang/DynamicProgramming/findTargetSumWays.go
package main

func findTargetSumWays(nums []int, target int) int {
	n := len(nums)

	summation := 0
	for _, v := range nums {
		summation += v
	}

	dp := make([][]int, n)
	for i := range dp {
		dp[i] = make([]int, 2*summation+1)
		for j := range dp[i] {
			dp[i][j] = -1
		}
	}

	var helper func(int, int) int
	helper = func(index int, sumNums int) int {
		if index < 0 {
			if sumNums == target {
				return 1
			}
			return 0
		}

		if dp[index][sumNums+summation] != -1 {
			return dp[index][sumNums+summation]
		}

		negative := helper(index-1, sumNums-nums[index])
		positive := helper(index-1, sumNums+nums[index])

		dp[index][sumNums+summation] = negative + positive
		return dp[index][sumNums+summation]
	}

	return helper(n-1, 0)
}
